### ISSUE TITLE
ORPO: the differentiated loss never touches the model (zero gradient), and orpo_loss is missing the paper's SFT term

### ISSUE THREAD

**Issue by swvgjbb78s-coder** [Goekdeniz-Guelmez/mlx-lm-lora#65]
Thanks for the quick fix on #55 — I confirmed the `importance_sampling_level` default is now
`"token"` on `main` and the token path returns `log_ratio` directly, so that one's resolved.

While working in the same area I hit two separate issues in the ORPO trainer. Both are still
present on `main` as of today. I've kept them in one issue since they're in the same file and
you'd likely touch both together, but they're independent defects.

## Bug 1 — `get_logps` is computed outside the differentiated function, so the gradient is identically zero

In `mlx_lm_lora/trainer/orpo_trainer.py`, `step()` builds the value-and-grad function at line
282, but calls `get_logps` at lines 288–289 — outside it — and then passes the resulting arrays
in as plain values at line 293:

```python
282:  loss_value_and_grad = nn.value_and_grad(model, loss_wrapper)
...
288:  chosen_logps, chosen_logits_mean = get_logps(model, chosen, chosen_masks)
289:  rejected_logps, rejected_logits_mean = get_logps(model, rejected, rejected_masks)
...
293:  (lvalue, reward, toks, metrics), grad = loss_value_and_grad(
294:      chosen_logps, chosen_logits_mean, rejected_logps, rejected_logits_mean, ...
```

Because every forward pass through `model` happens *before* `nn.value_and_grad` is invoked, the
function actually being differentiated is a pure function of already-materialised arrays. It has
no dependence on the model parameters, so the returned `grad` is zero with respect to them.

**What I observed.** Running ORPO at two learning rates for 35 iterations each, the adapter
weights were **bitwise unchanged** from initialisation, and the reported validation metrics were
identical to the last printed digit between iteration 1 and iteration 5. The run produces a
complete, plausible-looking training log while nothing is learned.

**Minimal fix.** Move the `get_logps` calls inside the function passed to `nn.value_and_grad`,
so the forward passes are part of the differentiated graph. (This is the same shape as the
correct pattern already used in `dpo_trainer.py`.)

## Bug 2 — `orpo_loss` omits the paper's NLL/SFT term

Lines 71–73:

```python
71:  log_odds = chosen_logps - rejected_logps
72:  ratio = nn.log_sigmoid(log_odds)
73:  loss = -beta * ratio
```

The ORPO objective in the paper (Hong et al., "ORPO: Monolithic Preference Optimization without
Reference Model") is a supervised NLL term on the chosen sequence *plus* the odds-ratio penalty.
Here only the relative margin remains. Without the NLL anchor the objective is degenerate: it can
be driven arbitrarily low by making the rejected sequence unlikely, with no pressure to keep the
chosen sequence well-formed.

**What I observed.** After fixing Bug 1 so gradients actually flowed, a 90-iteration run reached
validation accuracy 1.000 with margin +0.98 — while the model had **collapsed to emitting a single
token** by checkpoint 30. Accuracy 1.0 over a dead model; the metric was satisfied by the collapse.

**Minimal fix.** Add the SFT term back:

```python
nll_chosen = -chosen_logps          # length-averaged NLL of the chosen sequence
loss = nll_chosen - beta * ratio
```

## Notes

- Both were found on v2.1.0 and re-verified against `main` on 2026-07-18.
- I've been running with local patches for both; happy to open a PR if that's easier than a
  description, just say which you'd prefer.
- I have a third, weaker suspicion about `cpo_trainer.py` following a similar pattern, but I have
  **not** verified it and am deliberately not claiming it here.

**Comment by Goekdeniz-Guelmez** [comment]
Hey @swvgjbb78s-coder 

thanks for raising this issue, a patch is in the works and will the PR'ed later, feel free to look into it if you find something that can be made better. Do you have more in the cpo_trainer suspicion of yours?

**Comment by swvgjbb78s-coder** [comment]
Apologies for late response I've had a few things come up but Thanks — yes. I've now verified the cpo_trainer.py suspicion I flagged but deliberately
didn't claim in this issue. Same two defects as ORPO, plus two smaller ones in the metrics,
and one adjacent ORPO issue at the bottom. All still present on main (3.0.0, fb4f39db).
**1. Same zero-gradient shape as ORPO — but only on the default path.**
In step(), the forward pass happens before the differentiated call:
    policy_chosen_scores   = get_token_scores(model, chosen, chosen_masks)
    policy_rejected_scores = get_token_scores(model, rejected, rejected_masks)
    policy_chosen_score    = compute_score(policy_chosen_scores, chosen_masks, args.loss_type)
    policy_rejected_score  = compute_score(policy_rejected_scores, rejected_masks, args.loss_type)
    (lvalue, reward, toks, metrics), grad = loss_value_and_grad(
        policy_chosen_score, policy_rejected_score,
        chosen_masks=chosen_masks, rejected_masks=rejected_masks,
    )
loss_wrapper receives the scores as arguments and never references model, so
nn.value_and_grad(model, loss_wrapper) differentiates a function the model parameters
don't appear in. Same fix as ORPO's: have the wrapper take raw tokens and call
get_token_scores inside, exactly as dpo_trainer.py's loss_wrapper does.
**Worth flagging: seq_split_step() is correct — and that's the whole reason this hides.**
When seq_step_size is set it does a proper two-stage chain rule:
mx.grad(internal_loss_fn, argnums=[0, 1]) for dL/d(scores), then
mx.grad(local_loss_fn)(model) where local_loss_fn calls get_token_scores(model, chunk,
...) inside. But seq_step_size is only ever non-None when --efficient-long-context is
passed — train.py:725 for CPO, seq_step_size=512 if args.efficient_long_context else None
— and that flag is documented "Experimental." So CPO trains correctly only under an
experimental flag, and silently doesn't on the default documented path. That's probably
why it went unnoticed, and it gives you a working reference in the same file.
What I observed: a short CPO run — 5 iterations, --train-mode cpo --train-type lora
--num-layers 8 --learning-rate 1e-5 --batch-size 1, without --efficient-long-context so
the default step() path, resumed from an existing SFT adapter. The adapter written at end
of training is SHA-256 identical to the resume source: same hash, same 13,905,626 bytes,
all 112 tensors at max abs difference 0.0. CPO is reference-free so there's no KL term to
produce drift either — a zero policy gradient means literally zero movement, which is what
the bytes show.
Bounding that honestly: it's only 5 iterations, and the process aborted on a Metal OOM
right after the final save. So treat it as corroboration — the defect is provable by
reading step() and loss_wrapper, without running anything.
The check is the same one from #55:
    import mlx.core as mx
    a = mx.load("resume_source/adapters.safetensors")
    b = mx.load("cpo_output/adapters.safetensors")
    print(max(float(mx.abs(a[k].astype(mx.float32) - b[k].astype(mx.float32)).max()) for k in a))
    # 0.0
**2. cpo_loss is missing the NLL term — and for CPO that's the defining component.**
cpo_loss implements only the preference term (sigmoid/hinge/ipo/dpop over
policy_chosen_score - policy_rejected_score). Xu et al. define CPO as
L_CPO = L_prefer + L_NLL, where the NLL/behaviour-cloning anchor on the chosen sequence is
what distinguishes it from a reference-free DPO. Without it the objective is degenerate the
same way ORPO's was: nothing anchors absolute likelihood, only the gap.
**3. accuracies can only ever report 0.0 or 1.0.**
    chosen_reward   = beta * mx.mean(policy_chosen_score)   # scalar over the whole batch
    rejected_reward = beta * mx.mean(policy_rejected_score) # scalar
    "accuracies": mx.mean((chosen_reward > rejected_reward).astype(mx.float32)),
Both rewards are batch means, so that's scalar-vs-scalar — a single boolean, not a rate.
Per-example would be
mx.mean((policy_chosen_score > policy_rejected_score).astype(mx.float32)).
This matters more than it looks: it's the metric that would have surfaced (1) and (2).
**4. Asymmetric length normalisation in the reported logps.**
    "policy_rejected_logps": mx.mean(policy_rejected_score),                   # not normalised
    "policy_chosen_logps":   mx.mean(policy_chosen_score / num_chosen_tokens), # normalised
The two logged numbers aren't comparable — chosen reads far smaller than rejected for
reasons unrelated to the model. (chosen_logits_mean is also the un-normalised chosen score
again, under a "logits" label.) Cosmetic, but actively misleading while debugging the
first three.
**5. Adjacent, and ORPO-scoped so it belongs on this issue — but read, not run.**
In train.py's evaluate_model(), the ORPO branch (main:831-833) does:
    if args.train_mode == "orpo":
        efficient = args.seq_step_size is not None
        seq_step_size = args.seq_step_size or args.max_seq_length
There's no --seq-step-size argument and no seq_step_size key in the config defaults, so
unless a YAML config supplies it this should raise AttributeError on --test with
--train-mode orpo. I haven't executed this path, so treat it as a code read rather than a
reproduction.
Happy to open a PR covering (1)-(4). I had a look at #66 first — it touches
orpo_trainer.py and tests/test_training_algorithms.py only, so a CPO patch shouldn't
collide. If you'd rather fold CPO into your own pass, say so and I'll leave it alone.
*(Edited to correct a flag name: the working path is gated on --efficient-long-context,
not a --seq-step-size argument, which doesn't exist on the CLI. Also added (5).)*

**Comment by Goekdeniz-Guelmez** [comment]
If you want you can go ahead and open a PR

**Comment by Goekdeniz-Guelmez** [comment]
closed since its fixes are merged
